keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,360 @@
+  controller** `deployment`. Thus, when user invokes the **kube-migrator** tool,
+  the old `migration`s are deleted with the old `deployment`.
+
+The control loop of **kube-migrator controller** does the following:
+* watches the instances of the `migration` CR.
+* if `migration.spec.generation != migration.status.observedGeneration`
+    * deletes existing **migration worker** `job` for the resource type.
+    * launches new **migration worker** `job` to migrate the resource type. The
+      `job` is annotated with the generation.
+    * makes sure there is only one **migration worker** `job` running at a time,
+      to avoid overloading the apiserver.
+    * updates `migration.status.observedGeneration`.
+* watches the **migration worker** jobs. Updates
+  `migration.status.completedGeneration` when a job completes.
+* short-circuit: if `migration.spec.groupVersions` shows that the resource has
+  only one version, then its `.status.observedGeneration` and
+  `.status.completedGeneration` are updated directly without launching the
+  worker `job`.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in a `configmap`
+to preserve progress. With the inconsistent continue token introduced in
+[#67284][], the **migrator worker** does not need to worry about expired continue
+token.
+
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#api
+
+The cluster admin can run the `kube-migrator progress` command to check if the
+migration has completed. The command lists all `migrations` instances and checks
+if all of them have `.spec.generation == .status.observedGeneration`.
+
+### API
+
+We introduce the `migration` API to record the intention and progress of a
+migration. The API will be a CRD defined in the `migration.k8s.io` group.
+
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  Spec StorageVersionMigrationSpec
+  Status StorageVersionMigrationStatus
+}
+
+type StorageVersionMigrationSpec {
+  // Resource is the identifier of the resource that's being migrated.
+  // TBD: maybe resource=hash(etcd instance, etcd path prefix).
+  // Immuntable.
+  Resource string 
+  // GroupVersions are all the groups and versions this resource is served by
+  // the apiserver. The server preferred version should be groupVersions[0].
+  // **Migration workers** GET & PUT via the preferred version by default, to
+  // avoid conversions in the apiserver.
+  GroupVersions []GroupVersion
+  // Generation is the generation of the spec. Incrementing generation triggers
+  // migration. We do not use the objectMeta.generation because its semantic is
```

Comment:
Pretty sure we don't bump it for that.  If we do, we should flag it as not good and avoid doing that.

Reply:
Yeah, for CR only spec change bumps the generation ([code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L118-L126)). I'll update.

The issue I cited was about CRD actually. And FWIW, deployment increments generation for annotations changes, see [code](https://github.com/kubernetes/kubernetes/blob/58d43abc39178dc2efc3b1ed6e4727bb34a4b68a/pkg/registry/apps/deployment/strategy.go#L106).

Reply:
Updated the KEP to use the `metadata.generation`.